### PR TITLE
removing FastCGI support

### PR DESCRIPTION
**Pull request by aleksandra-tarkowska**
This PR removes FastCGI support (deprecated in https://github.com/openmicroscopy/openmicroscopy/pull/4056) in favor of WSGI as FastCGI support is deprecated since version 1.7 and will be removed in Django 1.9.

To resolve conflict with #4245 this PR replace https://github.com/openmicroscopy/openmicroscopy/pull/4259 

To test dependencies:
- unset ``omero.web.application_server``, uninstall Django ``pip uninstall Django`` and test ``bin/omero web help|start|stop|status|config|syncmedia|clearsessions|iis``. You should get error ``Django not installed!``
- install dependencies using ``pip install -r dist/requirements.txt`` (for python 2.6 use ``requirements_python26.txt``)
- uninstall gunicorn ``pip uninstall gunicorn`` and try export nginx config ``bin/omero web config nginx``, you should get errors ``ERROR: FastCGI support was removed in OMERO 5.2. Install Gunicorn...``
- set ``omero.web.application_server wsgi`` and try export apache config ``bin/omero web config apache``, no errors expected
- start web, statics should be copied and you should get message ``You are deploying OMERO.web using apache and mod_wsgi...``.
- install dependences using ``pip install -r dist/requirements.txt`` (for python 2.6 use ``requirements_python26.txt``)

To test config:
- set ``omero.web.application_server 'fastcgi-tcp'`` and test ``bin/omero web help|start|stop|status|config|syncmedia|clearsessions|iis``. You should get error ``Invalid APPLICATION_SERVER ...```
- run ``bin/omero admin diagnostics`` it should show you the same error in OMERO.web section.
- set ``omero.web.application_server development`` and try export nginx/apache config ``bin/omero web config nginx|apache``, you should get errors ``ERROR: configuration mismatch...``
- set ``omero.web.application_server wsgi`` and try export nginx config ``bin/omero web config nginx``, you should get errors ``ERROR: configuration mismatch...``
- unset ``omero.web.application_server`` and try export nginx config ``bin/omero web config apache``, you should get errors ``ERROR: configuration mismatch...``
- then try to export nginx config using old system flag `bin/omero web config nginx --system`` , you should get errors ``ERROR: --system is no longer supported...``

To test startup:
- unset ``omero.web.application_server`` and revert other settings to the defaults
- export nginx/apache config as in https://www.openmicroscopy.org/site/support/omero5.1/sysadmins/unix/install-web/install-wsgi.html and check if you do not get unexpected warnings errors, etc
- start web as normal ``bin/omero web start``
- check status  ``bin/omero web status``
- stop web  ``bin/omero web stop``

### CODE REVIEW

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#120](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/120/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/120/consoleText) for more details.
Possible conflicts:
  - Conflict detection failed
    - etc/templates/web/apache-fcgi.conf.template
    - etc/templates/web/apache-wsgi.conf.template
    - etc/templates/web/nginx-wsgi-development.conf.template
    - etc/templates/web/nginx-wsgi.conf.template
  - PR #4245 sbesson 'Configuration templates (part 2)'
    - etc/templates/web/nginx-development.conf.template
    - components/tools/OmeroPy/src/omero/plugins/web.py
    - etc/templates/nginx-wsgi-development.conf.template
    - etc/templates/web/apache.conf.template
    - etc/templates/nginx-wsgi.conf.template
    - etc/templates/apache-wsgi.conf.template
    - etc/templates/web/nginx.conf.template

**Comment by aleksandra-tarkowska** [comment]
Conflict was resolved, see description

**Comment by sbesson** [comment]
Note [OMERO-VIRTUALBOX-build](https://ci.openmicroscopy.org/job/OMERO-VIRTUALBOX-build/368/) is failing with

```
+ /home/omero/OMERO.server/bin/omero web config nginx-development --http 8080
ERROR: FastCGI support was removed in OMERO 5.2. Install Gunicorn and update config.
Build step 'Execute shell' marked build as failure
```

could you open a companion PR against ome-veewee to `pip install gunicorn` like https://github.com/ome/ome-veewee/pull/8/files?

**Comment by manics** [comment]
https://github.com/ome/ome-veewee is using ubuntu:14.04, so the distro verison (`apt-get install gunicorn`) should be sufficient.

**Comment by sbesson** [comment]
Also from https://ci.openmicroscopy.org/job/OMERO-DEV-merge-robotframework/127/console, the robot CLI plugin should be updated as well to not refer to `settings.FASTCGI_TYPES`

**Comment by aleksandra-tarkowska** [comment]
@sbesson the above commit https://github.com/aleksandra-tarkowska/openmicroscopy/commit/73429500ba1f4378177aa0d3d0e27a8792117e4c should resolve robot config issue

File: components/tests/ui/plugins/robot.py [hunk 1]
```diff
@@ -68,11 +68,13 @@ def config(self, args):
         d["WEBPREFIX"] = static_prefix
         d["QWEBPREFIX"] = urllib.quote(static_prefix, '')
         d["QSEP"] = urllib.quote('/', '')
-        if settings.APPLICATION_SERVER in settings.FASTCGI_TYPES:
-            d["WEBHOST"] = d["HOST"]
+        app_host = getattr(settings, 'APPLICATION_SERVER_HOST', None)
+        app_port = getattr(settings, 'APPLICATION_SERVER_PORT', None)
+        if app_host:
+            d["WEBHOST"] = ("%s:%s" %
+                            (app_host, app_port)) if app_port else app_host
```
Comment (sbesson) [inline components/tests/ui/plugins/robot.py]: See https://ci.openmicroscopy.org/job/OMERO-DEV-merge-robotframework/128/console, It looks like this defaulted to `127.0.0.1:4080`.

Comment (aleksandra-tarkowska) [inline components/tests/ui/plugins/robot.py]: As that is run on skate, why just not set application_server host and port properly. Previous logic doesn't have sense.

Comment (sbesson) [inline components/tests/ui/plugins/robot.py]: But we don't want the port here only the `application_server.host` - see https://ci.openmicroscopy.org/job/OMERO-DEV-merge-robotframework/129/console

Comment (aleksandra-tarkowska) [inline components/tests/ui/plugins/robot.py]: sorry, for some reason I had an impression they are empty by default. fixed now

**Comment by will-moore** [comment]
Trying to use removed options give appropriate response:
```
bin/omero web config apache-fcgi
usage: bin/omero web config [-h] [--http HTTP] [--max-body-size MAX_BODY_SIZE]
                            {nginx,nginx-development,apache}
bin/omero web config: error: argument type: invalid choice: 'apache-fcgi'

choose from:
	apache, nginx, nginx-development
```

**Comment by will-moore** [comment]
Not sure if this was the expected error:
```
$ bin/omero config set omero.web.application_server fastcgi-tcp
$ bin/omero web config nginx
Traceback (most recent call last):
  File "bin/omero", line 125, in <module>
    rv = omero.cli.argv()
  File "/Users/wmoore/Desktop/OMERO/dist/lib/python/omero/cli.py", line 1432, in argv
    cli.invoke(args[1:])
  File "/Users/wmoore/Desktop/OMERO/dist/lib/python/omero/cli.py", line 946, in invoke
    stop = self.onecmd(line, previous_args)
  File "/Users/wmoore/Desktop/OMERO/dist/lib/python/omero/cli.py", line 1023, in onecmd
    self.execute(line, previous_args)
  File "/Users/wmoore/Desktop/OMERO/dist/lib/python/omero/cli.py", line 1105, in execute
    args.func(args)
  File "/Users/wmoore/Desktop/OMERO/dist/lib/python/omero/plugins/web.py", line 233, in config
    from omeroweb import settings
  File "/Users/wmoore/Desktop/OMERO/dist/lib/python/omeroweb/settings.py", line 734, in <module>
    'DEPRECATED_SETTINGS_MAPPINGS')
  File "/Users/wmoore/Desktop/OMERO/dist/lib/python/omeroweb/settings.py", line 728, in process_custom_settings
    "Invalid %s JSON: %r" % (global_name, global_value))
ValueError: Invalid APPLICATION_SERVER JSON: 'fastcgi-tcp'
```

**Comment by aleksandra-tarkowska** [comment]
@will-moore I will wrap it in the more appropriate error message :-)

**Comment by aleksandra-tarkowska** [comment]
These are the final set of changes improving plugin and validation

File: components/tools/OmeroPy/src/omero/plugins/admin.py [hunk 2]
```diff
@@ -1398,13 +1398,14 @@ def port_val(port_type, value):
         from omero.plugins.web import WebControl
         try:
             WebControl().status(args)
-        except:
-            self.ctx.out("OMERO.web not installed!")
+        except Exception, e:
+            self.ctx.err("OMERO.web error:")
+            self.ctx.err(e)
```
Comment (aleksandra-tarkowska) [inline components/tools/OmeroPy/src/omero/plugins/admin.py]: @joshmoore how could I get only message from ctx.die like

```
OMERO.web error:
(682, ValueError('Invalid APPLICATION_SERVER (omero.web.application_ser...
```

Comment (sbesson) [inline components/tools/OmeroPy/src/omero/plugins/admin.py]: What's the output of `self.ctx.err(str(e))` ?

Comment (aleksandra-tarkowska) [inline components/tools/OmeroPy/src/omero/plugins/admin.py]: the same as above

Comment (sbesson) [inline components/tools/OmeroPy/src/omero/plugins/admin.py]: Discussed with @aleksandra-tarkowska, this seems to be catching an exception thrown by `self.ctx.die()` in which case a tuple is returned.

Comment (aleksandra-tarkowska) [inline components/tools/OmeroPy/src/omero/plugins/admin.py]: that is not a tuple, self.ctx.out("OMERO.web error: %r" % e.message[1].message)

**Comment by will-moore** [comment]
Without Django, get suitable error:
```
$ bin/omero web start
Django not installed!
```
Same for Gunicorn:
```
$ bin/omero web config nginx
ERROR: FastCGI support was removed in OMERO 5.2. Install Gunicorn and update config.
```
Other dependencies testing worked as expected.

Config: with fastcgi-tcp...
```
$ bin/omero web start
Invalid APPLICATION_SERVER (omero.web.application_server = 'fastcgi-tcp') ....
```
and same with ``` bin/omero admin diagnostics```.

With 'development' server, this works:
```
$ bin/omero web config nginx
ERROR: configuration mismatch. omero.web.application_server=development cannot be used with 'omero web config nginx'.
```
with 'wsgi', also get:
```
$ bin/omero web config nginx
ERROR: configuration mismatch. omero.web.application_server=wsgi cannot be used with 'omero web config nginx'.
```
and with no application server set:
```
$ bin/omero web config apache
ERROR: configuration mismatch. omero.web.application_server=wsgi-tcp cannot be used with 'omero web config apache'.
```
--system warning as expected:
```
$ bin/omero web config nginx --system
ERROR: --system is no longer supported, see --help
```
So far so good.....

**Comment by will-moore** [comment]
Started with...
```
$ omero config set omero.web.application_server
$ omero web config nginx > /usr/local/etc/nginx/others/omero_nginx.conf
# Edited to port 8080
$ omero web start
...
Starting OMERO.web... [OK]

# Webclient works fine...
$ bin/omero web status
OMERO.web status... [RUNNING] (PID 47311)

$ omero web stop
Stopping OMERO.web... [OK]
OMERO.web WSGI workers (PID 47311) killed.
```
All working as expected - good to merge.

**Comment by jburel** [comment]
Doc to be updated see https://trello.com/c/2Mzfzvho/74-developer-doc-updates

**Comment by jburel** [comment]
few "dependences" should be "dependencies"

**Comment by jburel** [comment]
The typo will be fixed in a follow up PR as doc PR.
Merging so it can be included in the build.

**Comment by jburel** [comment]
PR fixing typo now opened https://github.com/openmicroscopy/openmicroscopy/pull/4293
cc @aleksandra-tarkowska

**Pull request merged**